clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -436,4 +442,71 @@ static final String firstPrincipal(Subject subject) {
                 throw new KafkaException("Principal could not be determined from Subject, this may be a transient failure due to Kerberos re-login");
         }
     }
+
+    /**
+     * Kerberos exceptions that may require special handling. The standard Kerberos error codes
+     * for these errors are retrieved using KrbException#errorCode() from the underlying Kerberos
+     * exception thrown during {@link SaslClient#evaluateChallenge(byte[])}.
+     */
+    private enum KerberosError {
+        // (Mechanism level: Server not found in Kerberos database (7) - UNKNOWN_SERVER)
+        SERVER_NOT_FOUND(7, false),
+        // (Mechanism level: Ticket not yet valid (33) - Ticket not yet valid)])
+        // This could be a small timing window.
+        TICKET_NOT_YET_VALID(33, true),
+        // (Mechanism level: Request is a replay (34) - Request is a replay)
+        // Replay detection used to prevent DoS attacks can result in false positives, so retry on error.
+        REPLAY(34, true);
+
+
+        private static final Class<?> KRB_EXCEPTION_CLASS;
+        private static final Method KRB_EXCEPTION_RETURN_CODE_METHOD;
+
+        static {
+            try {
+                if (Java.isIbmJdk()) {
+                    KRB_EXCEPTION_CLASS = Class.forName("com.ibm.security.krb5.internal.KrbException");
+                } else {
+                    KRB_EXCEPTION_CLASS = Class.forName("sun.security.krb5.KrbException");
+                }
+                KRB_EXCEPTION_RETURN_CODE_METHOD = KRB_EXCEPTION_CLASS.getMethod("returnCode");
+            } catch (Exception e) {
+                throw new KafkaException("Kerberos exceptions could not be initialized", e);
```

Comment:
How will this manifest itself?

Reply:
More specifically, will a test fail if this stops working in a newer Java version?

Reply:
I have added one more test that checks the authentication failure case. That should reliably fail if we move to a newer Java version.